application.go

```diff
@@ -435,6 +442,20 @@ func (a *application) SetTools(args AgentToolsArgs) {
 	a.Tools_ = newAgentTools(args)
 }
 
+// CharmOrigin implements Application.
+func (a *application) CharmOrigin() CharmOrigin {
+	// To avoid a typed nil, check before returning.
+	if a.CharmOrigin_ == nil {
```

Comment:
Why would this be an issue? AFAICT from the concrete implementation, you would still be getting a NPE if you called `Source` on a nil `charmOrigin`.